### PR TITLE
dbft: implement fallback to multisignature block signing scheme

### LINKED ISSUE #352: Support fallback to multisignature block signing scheme
# Rationale

#347 introduces transition to threshold block signing scheme starting from NeoXAMEV fork. However, if something goes wrong with TPKE process, then dBFT should be able to fallback to the original multisignature scheme.

# Implementation

1. Include both TPKE shares and simple signatures data to the Commit message.
2. Include "fallback" NextConsensus value into block's Extra data: this field will be used to verify the set of multisignature validator public keys in case of fallback.
3. Include "reserved" NextConsensus into sealing block hash.
4. Check alignment of TPKE indexes from KeyManagement contract with multisignature validator public keys from Governance contract.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #352.

WIP, it's still in the process of testing. Consensus is running with both `--antimev.enforce-multisig-signing-scheme` enabled and disabled (checked for post-NeoXAMEV heights), RPC/Backps are also OK.

TODO:
- [X] Fix Extra construction.
- [x] Prettify the code and resolve minor TODOs.
- [x] Implement fallback from multisig to threshold scheme.
- [x] Ensure multisig scheme may be enforced at NeoXAMEV block.
- [x] Test the overall scheme.

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
@roman-khimov, review please, both `threshold -> multisig` and `multidig -> threshold` transitions are implemented and tested (if you'd like to test, then add `--antimev.enforce-multisig-signing-scheme` to the node's start command). Tthere's one edge-case question I doubt about (it's marked in the code).

Meanwhile, I'll resolve the rest of TODOs and rebase onto fresh master, we have some conflicts.

File: cmd/utils/flags.go [hunk 1]
```diff
@@ -513,6 +513,11 @@ var (
 		Usage:    "Password file to unlock anti-MEV keystore",
 		Category: flags.AntiMEVCategory,
 	}
+	AntiMEVEnforceMultisigBlockSigningSchemeFlag = &cli.BoolFlag{
+		Name:     "antimev.enforce-multisig-signing-scheme",
+		Usage:    "Enforce multidignature-based block signing scheme after NeoXAMEV transition with ExtraV1 block extra version",
```
Comment (roman-khimov) [inline cmd/utils/flags.go]: ECDSA-based, BLS is multisignature as well

File: cmd/utils/flags.go [hunk 2]
```diff
@@ -513,6 +513,11 @@ var (
 		Usage:    "Password file to unlock anti-MEV keystore",
 		Category: flags.AntiMEVCategory,
 	}
+	AntiMEVEnforceMultisigBlockSigningSchemeFlag = &cli.BoolFlag{
+		Name:     "antimev.enforce-multisig-signing-scheme",
```
Comment (roman-khimov) [inline cmd/utils/flags.go]: `force-ecdsa-signatures`

File: consensus/dbft/block.go [hunk 3]
```diff
@@ -108,13 +108,28 @@ func (b *Block) Verify(pub dbft.PublicKey, sign []byte) error {
 			return errors.New("invalid block signature")
 		}
 	case dbftutil.ExtraV1:
-		// We don't have a way to verify signature share, because the only way to
-		// verify is to collect at least M shares and verify *the group* of shares
-		// against *the global* public key. Hence, always consider Commit as valid
-		// at this stage.
-		return nil
+		switch sv := dbftutil.ExtraV1SigningScheme(b.header.Extra[1]); sv {
+		case dbftutil.ExtraV1MultisigScheme:
```
Comment (roman-khimov) [inline consensus/dbft/block.go]: s/multisig/ecdsa/

File: consensus/dbft/signer.go [hunk 4]
```diff
@@ -30,18 +30,25 @@ func (s *Signer) Sign(msg []byte) ([]byte, error) {
 	return s.SignFn(accounts.Account{Address: s.Signer}, accounts.MimetypeTextPlain, msg)
 }
 
-// signBlock signs block RLP bytes using the given signing scheme.
-func (s *Signer) signBlock(scheme dbftutil.ExtraVersion, blockRLP []byte) ([]byte, error) {
-	switch scheme {
+// signBlock signs block RLP bytes using the given extra version and signing scheme.
+func (s *Signer) signBlock(extra []byte, blockRLP []byte) ([]byte, error) {
```
Comment (roman-khimov) [inline consensus/dbft/signer.go]: type dbftExtra []byte

File: consensus/dbft/signer.go [hunk 5]
```diff
@@ -30,18 +30,25 @@ func (s *Signer) Sign(msg []byte) ([]byte, error) {
...
+	switch v := dbftutil.ExtraVersion(extra[0]); v {
```
Comment (roman-khimov) [inline consensus/dbft/signer.go]: extra.Version()

File: consensus/dbft/signer.go [hunk 6]
```diff
@@ -30,18 +30,25 @@ func (s *Signer) Sign(msg []byte) ([]byte, error) {
...
 	case dbftutil.ExtraV0:
 		return s.SignFn(accounts.Account{Address: s.Signer}, accounts.MimetypeTextPlain, blockRLP)
 	case dbftutil.ExtraV1:
-		share, err := s.AmevKeystore.SignShare(blockRLP)
-		if err != nil {
-			return nil, fmt.Errorf("failed to sign share: %w", err)
+		switch sv := dbftutil.ExtraV1SigningScheme(extra[1]); sv {
```
Comment (roman-khimov) [inline consensus/dbft/signer.go]: extra.SignatureScheme()

Can be a part of `dbftutil` and can be a separate commit. But it's too easy to make a mistake with 0/1/etc

File: consensus/dbft/dbft.go [hunk 7]
```diff
@@ -143,6 +147,10 @@ func getSignersAndSigsV0(cfg *config, extra []byte) ([]common.Address, [][]byte,
 		return nil, nil, fmt.Errorf("%w: expected %d, got %d", errUnexpectedExtraVersion, dbftutil.ExtraV0, extra[0])
 	}
 
+	return helper(cfg, extra[dbftutil.HashableExtraV0Len:])
+}
+
+func helper(cfg *config, src []byte) ([]common.Address, [][]byte, error) {
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: That's exactly the thing for this additional type.

File: consensus/dbft/dbft.go [hunk 8]
```diff
@@ -481,7 +503,19 @@ func New(chainCfg *params.ChainConfig, _ ethdb.Database) (*DBFT, error) {
 			ethBlock := finalBlock.ToEthBlock()
 			h := ethBlock.Header()
 			h.GasUsed = pre.finalGASUsed
-			h.MixDigest = c.getNextConsensus(h, pre.finalState)
+			multisig, threshold := c.getNextConsensus(h, pre.finalState)
+			// @roman-khimov, describe the NextConsensus setup behaviour for cases:
+			// 1. Primary scheme is used and we've got an error on threshold NextConsensus calculation
+			// 2. Fallback scheme is used, what should we put into threshold NextConsensus?
```
Comment (roman-khimov) [inline consensus/dbft/dbft.go]: We can have `NextConsensus` taking into account both schemes (it's a hash anyway). It'll always have ECDSA keys. It can have keys for threshold signatures. In any event you can check that the block is signed by one of these sets, but not anything else. Including only one set of keys doesn't allow to switch to me.

File: cmd/utils/flags.go (hunk 1 above)
Comment (AnnaShaleva) [reply cmd/utils/flags.go]: But threshold is not multisignature, and we don't use anything except ECDSA in dBFT.

Fixed.

File: cmd/utils/flags.go (hunk 2 above)
Comment (AnnaShaleva) [reply cmd/utils/flags.go]: Fixed.

File: consensus/dbft/block.go (hunk 3 above)
Comment (AnnaShaleva) [reply consensus/dbft/block.go]: Fixed.

File: consensus/dbft/signer.go (hunk 4 above)
Comment (AnnaShaleva) [reply consensus/dbft/signer.go]: Just `Extra`, we're at `dbftutil` package already.

File: consensus/dbft/signer.go (hunk 6 above)
Comment (AnnaShaleva) [reply consensus/dbft/signer.go]: Tried a separate commit, but then squashed, it simplifies a lot of things, so we don't need the old code.

File: consensus/dbft/dbft.go (hunk 8 above)
Comment (AnnaShaleva) [reply consensus/dbft/dbft.go]: Discussed, two NextConsensus are used.

**Comment by AnnaShaleva** [comment]
OK, I need to rebase now.

**Comment by AnnaShaleva** [comment]
OK, it works, but we still need https://github.com/bane-labs/go-ethereum/issues/353#issuecomment-2506294546, will be done in a separate PR.

File: eth/ethconfig/config.go [hunk 9]
```diff
@@ -161,6 +161,9 @@ type Config struct {
 
 	// OverrideVerkle (TODO: remove after the fork)
 	OverrideVerkle *uint64 `toml:",omitempty"`
+
+	// TODO: doc
```
Comment (roman-khimov) [inline eth/ethconfig/config.go]: TODO?

Comment (AnnaShaleva) [reply eth/ethconfig/config.go]: Documentation should be here, fixed.

**Review by roman-khimov — APPROVED** [review]
